### ISSUE TITLE
Global addDecorator is ignored

### ISSUE THREAD

**Issue by mjsarfatti** [storybookjs/storybook#10677]
**Describe the bug**
In my React/TypeScript app if I add a global `addDecorator` in my **preview.tsx** this is ignored.
The exact same decorator works if I add it to a single story.

What could it be?

**Expected behavior**
I expect the decorator to work in preview.tsx

**Code snippets**

**main.js**:

```js
const webpack = require("webpack");

// react-scripts-ts's webpack.config.js
const custom = require("react-scripts-ts/config/webpack.config.dev.js");

module.exports = {
  stories: ["../src/**/*.stories.(ts|tsx)"],
  addons: [
    "@storybook/addon-viewport",
    "@storybook/addon-storysource",
    "@storybook/addon-actions",
    "@storybook/addon-links",
  ],
  webpackFinal: (config) => {
    // "Finally, if your custom webpack config uses a loader that does not
    // explicitly include specific file extensions via the test property,
    // it is necessary to exclude the .ejs file extension from that loader."
    // https://storybook.js.org/docs/configurations/custom-webpack-config/#examples
    const oneOf = custom.module.rules[1].oneOf;
    oneOf[oneOf.length - 1].exclude.push(/\.ejs$/);

    const outConfig = {
      ...config,
      module: {
        ...config.module,
        rules: custom.module.rules,
      },
      plugins: [
        ...config.plugins,
        // See https://stackoverflow.com/a/51194917/416714 and
        // https://webpack.js.org/migrate/3/#loaderoptionsplugin-context
        new webpack.LoaderOptionsPlugin({
          options: {
            context: process.cwd(), // or the same value as `context`
          },
        }),
      ],
    };

    // Make sure we can import .ts/.tsx files in our stories
    outConfig.resolve.extensions.push('.ts', '.tsx');

    // Make sure storysource picks up .tsx source files
    outConfig.module.rules.push({
      test: /\.stories\.tsx?$/,
      loaders: [
        {
          loader: require.resolve('@storybook/source-loader'),
          options: { parser: 'typescript' },
        },
      ],
      enforce: 'pre',
    })

    return outConfig;
  },
};
```

**preview.tsx**:

```ts
import React from 'react';
import { addDecorator } from '@storybook/react';

// Doesn't do anything here
addDecorator((story) => <div style={{ background: 'pink' }}>{story()}</div>);
```

**button.stories.tsx**:

```ts
import React from 'react';
import { action } from '@storybook/addon-actions';
import Button from './button';

export default {
  title: 'UI Library/Controls/Buttons/Button',
  // The background actually turns pink here
  decorators: [(story) => <div style={{ background: 'pink' }}>{story()}</div>],
};

export const normal = () => (
  <Button onClick={action('clicked')}>Log In</Button>
);
```

**System:**

```
Environment Info:

  System:
    OS: macOS 10.15.4
    CPU: (8) x64 Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
  Binaries:
    Node: 12.13.1 - ~/.nvm/versions/node/v12.13.1/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.12.1 - ~/.nvm/versions/node/v12.13.1/bin/npm
  Browsers:
    Chrome: 81.0.4044.129
    Edge: 81.0.416.68
    Firefox: 76.0
    Safari: 13.1
  npmPackages:
    @storybook/addon-actions: ^5.3.18 => 5.3.18 
    @storybook/addon-links: ^5.3.18 => 5.3.18 
    @storybook/addon-storysource: ^5.3.18 => 5.3.18 
    @storybook/addon-viewport: ^5.3.18 => 5.3.18 
    @storybook/addons: ^5.3.18 => 5.3.18 
    @storybook/react: ^5.3.18 => 5.3.18 
```

**Comment by shilman** [comment]
i can assure you that global decorators work in storybook. do you have a public repro i can look at? if not @yannbf has created a great workflow to create one on codesandbox https://gist.github.com/yannbf/d7a32cfe697e71489db00bfd50e7ab5c

**Comment by brennop** [comment]
Was having the same problem. `preview.js` would not work at all. Using the latest beta fixed the problem for me

**Comment by shilman** [comment]
@brennop what version were you using before?

**Comment by mjsarfatti** [comment]
I'm using a custom webpack, basically replicating my project's config. For this reason (_I think_) I had to rename `preview.js` ☞ `preview.tsx` or it would error out trying to parse **jsx**.

Is it possible that this is causing problems? I have the feeling that `preview.tsx` is just ignored.

Will try to reproduce on a codesandbox in the meantime.

**Comment by mjsarfatti** [comment]
**_WORKAROUND_**

It appears that  
1. my current `react-scripts-ts` setup doesn't transform `.jsx` files, but only `.tsx` for some reason
2. Storybook ignores `preview.tsx` if it has a `.tsx` extension

I solved by 
1. renaming `preview.tsx` ☞ `preview.jsx`
2. adding the following line to my `main.js`:

```diff
[...]

  webpackFinal: (config) => {

    [...]

+   // Add Storybook's .jsx loader to the beginning of my custom webpack rules
+   custom.module.rules[1].oneOf.unshift(config.module.rules[0]);

    const outConfig = {
      ...config,
      module: {

      [...]
```

**Comment by shilman** [comment]
@mrmckeb is this something on our side?

**Comment by mrmckeb** [comment]
@shilman `react-scripts-ts` is a third-party scripts package, and is not actively maintained (two years since the last publish). Most people moved to `react-scripts` when we added native TypeScript support.

@mjsarfatti We have a good compatibility level between `react-scripts` and Storybook, I'd encourage you to move back to `react-scripts` if possible. 

Some of these other issues may be resolved in Storybook 6.0... which is now in beta.

**Comment by mjsarfatti** [comment]
@mrmckeb thank you for looking into this.

The problem is not so much with `react-scripts-ts` (which I'd be happy to move away from if it was under my control 😅), but with the fact that it appears that a file named `preview.tsx` is ignored, while if it is named `preview.jsx` it is not.

Would it be possible to add the `tsx` extension?

**Comment by mrmckeb** [comment]
This shouold be solved as a part of the below work @mjsarfatti. Have you tried Storybook 6 beta? If you can let me know, that would be great!
https://github.com/storybookjs/storybook/pull/10813/files#diff-b2967ea5bba766646d93fe3653239d67R155

**Comment by mjsarfatti** [comment]
Haven't tried it yet, but it seems like it should solve the issue :)

**Comment by taisuke-j** [comment]
Yes, `preview.tsx` does seem to work now.

**Comment by fabiosantoscode** [comment]
I had an issue with a context provider, which others might have as well. I see lots of this in Storybook examples:

```
story => <Provider>{story()}</Provider>
```

So that the Provider can provide its context, I had to replace it with:

```
Story => <Provider><Story /></Provider>
```

And it's obvious why I had to do it, but I feel dumb because of how much it took for me to realize it.

**Comment by y-nk** [comment]
This also happens to me now (v6.3.0) with a simple javascript (no typescript) project.

**Comment by shilman** [comment]
@y-nk do you have a repro?

**Comment by y-nk** [comment]
@shilman no, but i found the root cause - at least for my own case and it was pretty annoying to discover about it.

```
npx create-react-app my-app
cd my-app
npx sb init
```

```
// .storybook/preview.js

import React from 'react';
import { addDecorator } from '@storybook/react';

console.log("preview.js")
addDecorator(Story => {
  console.log("in decorator")
  return <Story />
})
```

With just this, run `yarn storybook` and (at least for me) the console.log "in decorator" would not show up... until... i clicked on a ".js" story.

I spent hours trying to go around the issue (literally reverting until 6.0.0), but it appears the decorator wasn't loading for the default opening `.mdx` story.

I could work a minimal repro in codesandbox if you need @shilman let me know.

**Comment by shilman** [comment]
@y-nk Thanks I'll take a look and let you know 🙏

**Comment by dko-slapdash** [comment]
@shilman Seems like I experience the similar issue, preview.js decorators just don’t apply (except ~5% of cases when they do, but I can’t reproduce the success, it appears randomly and disappears after the page reload).

**Comment by shilman** [comment]
@dko-slapdash @y-nk can one of you provide a repro by running `npx sb@next repro`, following the instructions, and linking it here? 🙏

**Comment by dko-slapdash** [comment]
@shilman I actually found a work-around (maybe not even a work-around, but a proper solution). In the previous version, we had the following code in preview.jsx:
```
configure(
  require.context("client/dist", true, /\.stories\.js$/),
  module
);
```

And this code caused the above effect of ignoring decorators in 95% of case. Once I removed it and instead added `stories: ["../../client/dist/**/*.stories.js"]` to main.ts, it started working as expected. (Here `client` is the second yarn workspaces monorepo project in packages/.)
